File: api/TotemAPI/TotemAPI/Features/Checkout/Infrastructure/FakeTefPaymentService.cs
using TotemAPI.Features.Checkout.Application.Abstractions;
using TotemAPI.Features.Checkout.Domain;

namespace TotemAPI.Features.Checkout.Infrastructure;

public sealed class FakeTefPaymentService : ITefPaymentService
{
    public Task<TefPixCharge> CreatePixChargeAsync(int amountCents, string reference, CancellationToken ct)
    {
        if (amountCents <= 0) throw new ArgumentException("amountCents inválido.");
        var now = DateTimeOffset.UtcNow;
        var expiresAt = now.AddMinutes(5);
        var providerReference = $"pix-{Guid.NewGuid():N}";
        var payload = $"000201|AMOUNT={amountCents}|REF={reference}|EXPIRES={expiresAt:O}|PR={providerReference}";
        return Task.FromResult(new TefPixCharge(payload, expiresAt, providerReference));
    }

    public Task<string> StartCardAsync(int amountCents, PaymentMethod method, string reference, CancellationToken ct)
    {
        if (amountCents <= 0) throw new ArgumentException("amountCents inválido.");
        if (method is not PaymentMethod.CreditCard and not PaymentMethod.DebitCard) throw new ArgumentException("method inválido.");
        var providerReference = $"card-{Guid.NewGuid():N}";
        return Task.FromResult(providerReference);
    }

    public Task<TefPaymentConfirmation> ConfirmAsync(string providerReference, PaymentMethod method, CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(providerReference)) throw new ArgumentException("providerReference inválido.");
        var transactionId = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds().ToString();
        return Task.FromResult(new TefPaymentConfirmation(true, transactionId, "APROVADO"));
    }
}

File: lib/src/features/checkout/data/services/fake_tef_payment_service.dart
import '../../domain/entities/checkout_order.dart';
import '../../domain/entities/payment_result.dart';
import '../../domain/entities/pix_charge.dart';
import '../../domain/services/payment_service.dart';

class FakeTefPaymentService implements PaymentService {
  const FakeTefPaymentService();

  @override
  Future<PixCharge> createPixCharge({
    required int amountCents,
    required String reference,
  }) async {
    await Future<void>.delayed(const Duration(milliseconds: 400));
    final now = DateTime.now();
    final expiresAt = now.add(const Duration(minutes: 5));
    final payload = '000201|AMOUNT=$amountCents|REF=$reference|EXPIRES=${expiresAt.toIso8601String()}';
    return PixCharge(
      amountCents: amountCents,
      payload: payload,
      expiresAt: expiresAt,
      reference: reference,
    );
  }

  @override
  Future<PaymentResult> pay({
    required int amountCents,
    required PaymentMethod method,
  }) async {
    final delay = switch (method) {
      PaymentMethod.creditCard || PaymentMethod.debitCard => const Duration(seconds: 30),
      PaymentMethod.pix => const Duration(milliseconds: 700),
      PaymentMethod.cash => const Duration(milliseconds: 200),
    };
    await Future<void>.delayed(delay);
    return PaymentResult(
      isApproved: true,
      transactionId: DateTime.now().millisecondsSinceEpoch.toString(),
      message: 'APROVADO',
    );
  }
}
